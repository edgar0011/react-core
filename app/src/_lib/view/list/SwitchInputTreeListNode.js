// @flow

import React, { PureComponent } from 'react'
import type { Node } from 'react'
import type { HeadlessRender } from '../common/HeadlessComponentBase'
import type { OptionTypesWithValue, ValuesType } from './SwitchInput'

import SwitchInputList from './SwitchInputList'

type SwitchInputTreeListNodeProps = {
  key: string | number,
  nodeLabel: string,
  options: ?OptionTypesWithValue,
  value?: ValuesType,
  components: {
    option: HeadlessRender,
    select: HeadlessRender
  },
  onChange: ({ target: { value: ?any }}) => void
}
/**
 * @class
 * @name SwitchInputTreeListNode
 * @classdesc Switch input tree list node.
 * @extends PureComponent
 *
 * @prop {function} render Render function must be specified unless RenderComponent is specified.
 * @prop {ComponentType} RenderComponent Render component must be specified unless render function is specified.
 * @prop {string} nodeLabel Node label text.
 * @prop {Object} components Render configuration container.
 * @prop {Object} components.option Option render configuration container.
 * @prop {function} components.option.render Option render function.
 * @prop {any} components.option.RenderComponent Option render component.
 * @prop {Object} components.select Select render configuration container.
 * @prop {function} components.select.render Select render function.
 * @prop {any} components.select.RenderComponent Select render component.
 */
export default class SwitchInputTreeListNode extends PureComponent<SwitchInputTreeListNodeProps> {

  constructor(props: SwitchInputTreeListNodeProps) {
    super(props)
  }

  props: SwitchInputTreeListNodeProps

  /**
   * Click handler for switching selected options.
   * @memberOf SwitchInputTreeListNode
   * @method SwitchInputTreeListNode#clickHandler
   * @param {*} selected Selected flag.
   */
  clickHandler = (selected: boolean) => {
    if (this.props.options) {
      this.props.onChange({ target: { value: selected ? [] : this.props.options.map(opt => opt.value) } })
    }
  }

  /**
   * Component render function.
   * @memberOf SwitchInputTreeListNode
   * @method SwitchInputTreeListNode#render
   * @returns {*}
   */
  render() {
    const {
      nodeLabel, value, options, components: { select: { render, RenderComponent } }, ...props
    }: SwitchInputTreeListNodeProps = this.props

    const allSelected = options && value ? value.length === options.length : false

    let renderedSelect: Node
    const selectProps = {
      bordered: true,
      selected: allSelected,
      label: nodeLabel,
      onClick: this.clickHandler,
      style: { margin: '10px', marginRight: '0', padding: '5px' }
    }
    if (render) {
      renderedSelect = render(selectProps)
    } else {
      renderedSelect = RenderComponent ? <RenderComponent {...selectProps} /> : null
    }

    return (
      <div style={{ width: '100%' }}>
        <div>
          {renderedSelect}
        </div>
        <div style={{ paddingLeft: '20px' }}>
          <SwitchInputList
            {...props}
            multiple
            options={options}
            value={value}
          />
        </div>
      </div>
    )
  }
}
