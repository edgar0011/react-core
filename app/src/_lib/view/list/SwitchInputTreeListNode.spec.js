import React from 'react'
import SwitchInputTreeListNode from './SwitchInputTreeListNode'

describe('SwitchInputTreeListNode', () => {
  it('should render SwitchInputTreeListNode', () => {
    const props = {
      components: {
        option: {},
        select: {},
      }
    }
    global.shallowSnapshot(<SwitchInputTreeListNode {...props} />)
  })
})
