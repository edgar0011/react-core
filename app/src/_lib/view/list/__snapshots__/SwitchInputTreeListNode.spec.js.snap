// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchInputTreeListNode should render SwitchInputTreeListNode 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SwitchInputTreeListNode
    components={
      Object {
        "option": Object {},
        "select": Object {},
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div />,
        <div
          style={
            Object {
              "paddingLeft": "20px",
            }
          }
        >
          <SwitchInputList
            multiple={true}
            options={undefined}
            value={undefined}
          />
        </div>,
      ],
      "style": Object {
        "width": "100%",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <SwitchInputList
            multiple={true}
            options={undefined}
            value={undefined}
          />,
          "style": Object {
            "paddingLeft": "20px",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "multiple": true,
            "options": undefined,
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div />,
          <div
            style={
              Object {
                "paddingLeft": "20px",
              }
            }
          >
            <SwitchInputList
              multiple={true}
              options={undefined}
              value={undefined}
            />
          </div>,
        ],
        "style": Object {
          "width": "100%",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <SwitchInputList
              multiple={true}
              options={undefined}
              value={undefined}
            />,
            "style": Object {
              "paddingLeft": "20px",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "multiple": true,
              "options": undefined,
              "value": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
